server/lib/object-storage.ts

```diff
@@ -0,0 +1,176 @@
+import {
+  CompletedPart,
+  CompleteMultipartUploadCommand,
+  CreateMultipartUploadCommand,
+  DeleteObjectCommand,
+  DeleteObjectsCommand,
+  GetObjectCommand,
+  ListObjectsV2Command,
+  PutObjectCommand,
+  S3Client,
+  UploadPartCommand
+} from "@aws-sdk/client-s3"
+import { CONFIG } from "@server/initializers/config"
+import { logger } from '@server/helpers/logger'
+import { createReadStream, createWriteStream, ensureDir, open, close, ReadStream, stat, Stats, remove } from "fs-extra"
+import { Readable } from "stream"
+import { dirname } from "path"
+import { min } from "lodash"
+import { pipelinePromise } from "@server/helpers/core-utils"
+
+type BucketInfo = {BUCKET_NAME: string, PREFIX?: string, BASE_URL?: string}
+const ONE_MIB = 1024 * 1024
+const MAX_PUT_SIZE = process.env.NODE_ENV.includes("test") ? 10 * ONE_MIB : 100 * ONE_MIB
+
+function getS3Client () {
+  return new S3Client({ endpoint: CONFIG.OBJECT_STORAGE.ENDPOINT.toString() })
+}
+
+function getPartSize (stats: Stats) {
+  if (process.env.NODE_ENV.includes("test")) {
+    return 10 * ONE_MIB
+  }
+  // Use parts of 1 GiB if the file is very large (it would take more than 1000 requests at 100 MiB per request)
```

Comment:
I don't think this is needed. 
The part size should be as large as possible, but not causing timeout errors. It is better to leave the size at 2GB or add it to the config as `мultipartTriggerSize`, for example.

Reply:
What do you mean "leave the size at 2GB"? 

I could add a configuration option, but I think these are fine defaults. Backblaze b2 recommends 100MB in `recommendedPartSize` (see [docs](https://www.backblaze.com/b2/docs/large_files.html), so I think that's a good default (although I use 100MiB here, instead of MB). The only reason this function uses 1GiB sometimes is because otherwise it isn't possible to upload files larger than 1TB.

Reply:
> We recommend a part size of 100MB, which strikes a good balance between upload throughput and the ability to **upload parts in parallel**.
No paralel parts uploading is used here, so a larger size is preferable for better performance with less pay.

Reply:
That's a good point